jib-core/src/main/java/com/google/cloud/tools/jib/api/JavaContainerBuilder.java

```diff
@@ -113,10 +106,64 @@ public static JavaContainerBuilder from(RegistryImage registryImage) {
   private final List<String> jvmFlags = new ArrayList<>();
   private final LinkedHashSet<String> classpath = new LinkedHashSet<>(4);
 
+  /** Absolute path of directory containing application resources on container. */
+  private AbsoluteUnixPath resourcesPath;
+
+  /** Absolute path of directory containing classes on container. */
+  private AbsoluteUnixPath classesPath;
+
+  /** Absolute path of directory containing dependencies on container. */
+  private AbsoluteUnixPath dependenciesPath;
+
+  /** The entrypoint classpath element corresponding to dependencies. */
+  private AbsoluteUnixPath dependenciesClasspath;
+
+  /** Absolute path of directory containing additional classpath files on container. */
+  private AbsoluteUnixPath othersPath;
+
   @Nullable private String mainClass;
 
   private JavaContainerBuilder(JibContainerBuilder jibContainerBuilder) {
     this.jibContainerBuilder = jibContainerBuilder;
+    AbsoluteUnixPath appRoot = AbsoluteUnixPath.get("/app");
+    resourcesPath =
+        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_RESOURCES_PATH_ON_IMAGE);
+    classesPath = appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_CLASSES_PATH_ON_IMAGE);
+    dependenciesPath =
+        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_DEPENDENCIES_PATH_ON_IMAGE);
+    dependenciesClasspath = dependenciesPath.resolve("*");
+    othersPath = appRoot.resolve("classpath");
+  }
+
+  /**
+   * Sets the app root of the container image (useful for building WAR containers).
+   *
+   * @param appRoot the absolute path of the app on the container ({@code /app} by default)
+   * @return this
+   */
+  public JavaContainerBuilder setAppRoot(String appRoot) {
+    return setAppRoot(AbsoluteUnixPath.get(appRoot));
+  }
+
+  /**
+   * Sets the app root of the container image (useful for building WAR containers).
+   *
+   * @param appRoot the absolute path of the app on the container ({@code /app} by default)
+   * @return this
+   */
+  public JavaContainerBuilder setAppRoot(AbsoluteUnixPath appRoot) {
+    if (!classpath.isEmpty()) {
```

Comment:
Looking at the exception message "App root must be set before adding files", I don't really get this check. We are given `appRoot` by the parameter and setting this now and this methods doesn't add files.

Reply:
Do you mean the exception message is unclear, or you don't understand why we're doing the check in the first place? The check is there because right now, setting the app root doesn't change the app root for previously added files. e.g. if you add class files, change app root, then add dependencies, now the dependencies are in a different root directory than the classes that were added earlier. I added the check so people don't assume calling `setAppRoot` updates previously added layers.

Reply:
Hmm, we might want to make this declarative and be able to allow the user to set the app root whenever, including after adding the layers.

Reply:
Ah, now I get it. I think the error message is a bit confusing. Something like "you cannot set a new app root (or change the app root) after files are added" seems clearer.

Reply:
@coollog I think I'll just file an issue for that since it seems like it'd involve changing a lot of internals of the class.